README.md
[![Actions Status](https://github.com/maxmilhas/ts-base-http-client/workflows/build/badge.svg)](https://github.com/maxmilhas/ts-base-http-client/actions)
[![Actions Status](https://github.com/maxmilhas/ts-base-http-client/workflows/test/badge.svg)](https://github.com/maxmilhas/ts-base-http-client/actions)
[![Actions Status](https://github.com/maxmilhas/ts-base-http-client/workflows/lint/badge.svg)](https://github.com/maxmilhas/ts-base-http-client/actions)
[![Test Coverage](https://api.codeclimate.com/v1/badges/65e41e3018643f28168e/test_coverage)](https://codeclimate.com/github/maxmilhas/ts-base-http-client/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/65e41e3018643f28168e/maintainability)](https://codeclimate.com/github/maxmilhas/ts-base-http-client/maintainability)
[![Packages](https://david-dm.org/maxmilhas/ts-base-http-client.svg)](https://david-dm.org/maxmilhas/ts-base-http-client)
[![npm version](https://badge.fury.io/js/%40maxmilhas%2Fts-base-http-client.svg)](https://badge.fury.io/js/%40maxmilhas%2Fts-base-http-client)

This project offers a base http client with standardized parameters, ready to be extended or injected

## Quick Start

* To get this project running in your machine, first, install all dependencies `npm install`;
* Then, you need to create a *.env* file to set it up. You can use *.env.staging* or *.env.test* as an example;
* Finally, you can run `npm run start:dev` do test it directly running the .ts files;
* If you want to build it, run `npm run build`
* You can also run unit tests with coverage, with `npm run test:coverage`

### Who do I talk to

This project is maintained by the

## Requirements

To guarantee the correct functioning of this package the following dependencies are required:

* Node v10.0 or higher
* TypeScript v3.0 or higher

## How to use

This injecting purposes, BaseHttpClient can be extending and the constructor parameters passed with the corresponding injection tokens for your injection library, example:

``` ts
@Injectable()
export class MyHttpClient extends BaseHttpClient {
	constructor(
		config: MyApiConfig,
		logger: Logger,
	) {
		super(config, logger);
	}
}
```

MyApiConfig must implements [ApiConfig](./src/api-config.ts). The following properties are optional, though:
```ts
retries?: number; // default 0
logRequests?: boolean; // default false
contentType?: string; // default application/json
```

As you see, you also need to provide a Logger compatible with the [Logger interface](./src/Logger.ts).
Is you set logRequests to true, all the requests made by that instance of client will be logged i the curl format, which is very useful for debug purposes (but not quite production fit).


The methods **get**, **post**, **put**, **delete** and **patch** have a closed use and will return directly the body of the request. If you need more advanced use, rely on **req**, a method where you need to specify the verb of your http request as the first parameters, but it returns a known superagent request object with timeout, retry, content-type and, if enabled, the logger, already set,

## References

* [www.maxmilhas.com.br](https://www.maxmilhas.com.br/ "www.maxmilhas.com.br")

## Contributing

Anyone from MaxMilhas can contribute to this project since the changes follows the rules described on [CONTRIBUTING.md](./CONTRIBUTING.md).


CHANGELOG.md
## [0.2.3](https://github.com/maxmilhas/ts-base-http-client/compare/v0.2.2...v0.2.3) (2023-07-31)


### Bug Fixes

* fixing publishing ([1aeb53a](https://github.com/maxmilhas/ts-base-http-client/commit/1aeb53acd3352261d4f6f061e6a4363ab09902b2))

## [0.2.2](https://github.com/maxmilhas/ts-base-http-client/compare/v0.2.1...v0.2.2) (2023-07-31)


### Bug Fixes

* fixing publishing ([0d16927](https://github.com/maxmilhas/ts-base-http-client/commit/0d1692777785b9d4ab3acb6ec301e57c3df023ed))

## [0.2.1](https://github.com/maxmilhas/ts-base-http-client/compare/v0.2.0...v0.2.1) (2023-07-31)


### Bug Fixes

* fixing publishing ([43899e9](https://github.com/maxmilhas/ts-base-http-client/commit/43899e97e07d1c903b16e55a010bede0299f6e1d))

# [0.2.0](https://github.com/maxmilhas/ts-base-http-client/compare/v0.1.1...v0.2.0) (2023-07-31)


### Features

* adding getErrorMessage helper ([e302d6e](https://github.com/maxmilhas/ts-base-http-client/commit/e302d6e0adcf32eee45ecdd7bdda7606dc9372e0))

## [0.1.1](https://github.com/maxmilhas/ts-base-http-client/compare/v0.1.0...v0.1.1) (2023-01-04)


### Bug Fixes

* fixing wrong default retry ([bc0fd71](https://github.com/maxmilhas/ts-base-http-client/commit/bc0fd71d355fb1f9a083677a65d461f6936fcb3a))

# [0.1.0](https://github.com/maxmilhas/ts-base-http-client/compare/v0.0.0...v0.1.0) (2022-07-12)


### Features

* first version ([7a54765](https://github.com/maxmilhas/ts-base-http-client/commit/7a547658a581489ae2d6755e7a4385b2821f9ced))


test/unit/get-api-config.spec.ts
import { ApiConfig, getApiConfig } from './../../src';
import ms = require('ms');

describe(getApiConfig.name, () => {
	it('should return all the default values for the non informed envs vars', () => {
		const envs: Record<string, string | undefined> = {
			MY_API_URL: 'my url',
		};

		const result = getApiConfig(envs, 'MY_API');

		expect(result).toEqual({
			timeout: ms('30s'),
			url: 'my url',
			retries: undefined,
			contentType: undefined,
			logRequests: false,
		} as ApiConfig);
	});

	it('should return all the informed default values for the non informed envs vars', () => {
		const envs: Record<string, string | undefined> = {
			MY_API_BASE_URI: 'my url',
		};

		const result = getApiConfig(envs, 'MY_API', {
			contentType: 'application/pokemon',
			retries: 10,
			timeout: ms('3s'),
		});

		expect(result).toEqual({
			timeout: ms('3s'),
			url: 'my url',
			retries: 10,
			contentType: 'application/pokemon',
			logRequests: false,
		} as ApiConfig);
	});

	it('should return all the informed values', () => {
		const envs: Record<string, string | undefined> = {
			MY_API_BASE_URL: 'my url',
			MY_API_TIMEOUT: '10s',
			MY_API_RETRIES: '5',
			MY_API_LOG_REQUESTS: 'true',
			MY_API_CONTENT_TYPE: 'application/he-man',
		};

		const result = getApiConfig(envs, 'MY_API', {
			retries: 10,
			timeout: ms('3s'),
		});

		expect(result).toEqual({
			timeout: ms('10s'),
			url: 'my url',
			retries: 5,
			contentType: 'application/he-man',
			logRequests: true,
		} as ApiConfig);
	});
});


test/unit/base-http-client.spec.ts
let superdebug: jest.SpyInstance;
let agentKeepAlive: jest.SpyInstance;
jest.mock(
	'superdebug',
	() => (superdebug = jest.fn().mockReturnValue('superdebug engaged')),
);
jest.mock(
	'agentkeepalive',
	() => (agentKeepAlive = jest.fn().mockImplementation((x) => x)),
);
import AgentKeepAlive = require('agentkeepalive');
import superagent = require('superagent');
import { ApiConfig, BaseHttpClient, Methods, Logger } from '../../src';
import { expectCallsLike, getNames } from '../jest-setup';

const methods = getNames(BaseHttpClient);

describe(BaseHttpClient.name, () => {
	let target: BaseHttpClient;
	let config: ApiConfig;
	let logger: Logger;

	beforeEach(() => {
		config = { url: '' } as any;
		logger = {} as any;

		target = new BaseHttpClient(config, logger);
	});

	it('should start correctly with https', () => {
		const stub = ((agentKeepAlive as any).HttpsAgent = jest
			.fn()
			.mockImplementation((x) => x));

		const result = new BaseHttpClient({ url: 'hTtpS://aaa' } as any, logger);

		expect(stub).toHaveBeenCalled();
		expect(result).toBeInstanceOf(BaseHttpClient);
	});

	it('should assume application/json as default contentType when it is not informed', () => {
		expect(target['defaultContentType']).toBe('application/json');
	});

	it('should assume default Agent parameters when no agent config is informed', () => {
		expectCallsLike(agentKeepAlive, [
			{
				maxFreeSockets: 50,
				timeout: 60000,
				freeSocketTimeout: 30000,
			},
		]);
	});

	it('should use informed contentType', () => {
		config.contentType = 'new content type';
		const newTarget = new BaseHttpClient(config, logger);

		expect(newTarget['defaultContentType']).toBe('new content type');
	});

	it('should use informed agent config', () => {
		config.agent = { info: 'new content type' } as any;
		new BaseHttpClient(config, logger);

		expect(agentKeepAlive).toHaveBeenCalledWith({
			info: 'new content type',
		});
	});

	describe(methods.req, () => {
		let request: superagent.SuperAgentRequest;
		let mockedHttpMethod: jest.SpyInstance;

		beforeEach(() => {
			config.retries = 'retries value' as any;
			config.timeout = 'timeout value' as any;
			config.url = 'url value' as any;
			request = Promise.resolve('expected result') as any;
			request.retry = jest.fn().mockReturnValue(request);
			request.timeout = jest.fn().mockReturnValue(request);
			request.set = jest.fn().mockReturnValue(request);
			request.use = jest.fn().mockReturnValue(request);
			request.agent = jest.fn().mockReturnValue(request);
			mockedHttpMethod = (superagent as any).someHttpMethod = jest
				.fn()
				.mockReturnValue(request);
			logger.info = { bind: jest.fn().mockReturnValue('bound info') } as any;
		});

		it('should call the resource with the given method and the configured timeout and number of retries', async () => {
			const result = await target.req(
				'someHttpMethod' as any,
				'the resource that I want',
				'my content type',
			);

			expectCallsLike(mockedHttpMethod, ['url value/the resource that I want']);
			expectCallsLike(request.retry, ['retries value']);
			expectCallsLike(request.timeout, ['timeout value']);
			expectCallsLike(request.set, ['content-type', 'my content type']);
			expectCallsLike(request.agent, [
				new AgentKeepAlive({
					maxFreeSockets: 50,
					timeout: 60000,
					freeSocketTimeout: 30000,
				}),
			]);
			expect(superdebug).toBeUndefined();
			expectCallsLike(request.use);
			expectCallsLike(logger.info.bind);
			expect(result).toEqual('expected result');
		});

		it('should call the resource with the given method and the configured timeout and no retries and default content type, when they are not informed', async () => {
			(target as any).defaultContentType = 'default content type';
			config.retries = undefined;

			const result = await target.req(
				'someHttpMethod' as any,
				'the resource that I want',
			);

			expectCallsLike(mockedHttpMethod, ['url value/the resource that I want']);
			expectCallsLike(request.retry, [0]);
			expectCallsLike(request.timeout, ['timeout value']);
			expectCallsLike(request.set, ['content-type', 'default content type']);
			expect(superdebug).toBeUndefined();
			expectCallsLike(request.use);
			expectCallsLike(request.agent, [
				new AgentKeepAlive({
					maxFreeSockets: 50,
					timeout: 60000,
					freeSocketTimeout: 30000,
				}),
			]);
			expectCallsLike(logger.info.bind);
			expect(result).toEqual('expected result');
		});

		it('should call the resource and log the call when logRequests is true', async () => {
			config.logRequests = true;

			const result = await target.req(
				'someHttpMethod' as any,
				'the resource that I want',
				'my content type',
			);

			expectCallsLike(mockedHttpMethod, ['url value/the resource that I want']);
			expectCallsLike(request.retry, ['retries value']);
			expectCallsLike(request.timeout, ['timeout value']);
			expectCallsLike(request.set, ['content-type', 'my content type']);
			expectCallsLike(logger.info.bind, [logger]);
			expectCallsLike(superdebug, ['bound info']);
			expectCallsLike(request.use, ['superdebug engaged']);
			expectCallsLike(request.agent, [
				new AgentKeepAlive({
					maxFreeSockets: 50,
					timeout: 60000,
					freeSocketTimeout: 30000,
				}),
			]);
			expect(result).toEqual('expected result');
		});
	});

	describe(methods.post, () => {
		let request: superagent.SuperAgentRequest;

		beforeEach(() => {
			request = Promise.resolve({ body: 'expected result' }) as any;
			request.send = jest.fn().mockReturnValue(request);
			jest.spyOn(target, 'req').mockReturnValue(request);
		});

		it('should call the resource with the given method and the configured timeout and number of retries', async () => {
			const result = await target.post(
				'the resource that I want',
				{
					payload: 'value',
				},
				'my content type',
			);

			expectCallsLike(target.req, [
				Methods.POST,
				'the resource that I want',
				'my content type',
			]);
			expectCallsLike(request.send, [
				{
					payload: 'value',
				},
			]);
			expect(result).toEqual('expected result');
		});
	});

	describe(methods.delete, () => {
		let request: superagent.SuperAgentRequest;

		beforeEach(() => {
			request = Promise.resolve({ body: 'expected result' }) as any;
			jest.spyOn(target, 'req').mockReturnValue(request);
		});

		it('should call the resource with the given method and the configured timeout and number of retries', async () => {
			const result = await target.delete(
				'the resource that I want',
				'my content type',
			);

			expectCallsLike(target.req, [
				Methods.DELETE,
				'the resource that I want',
				'my content type',
			]);
			expect(result).toEqual('expected result');
		});
	});

	describe(methods.patch, () => {
		let request: superagent.SuperAgentRequest;

		beforeEach(() => {
			request = Promise.resolve({ body: 'expected result' }) as any;
			request.send = jest.fn().mockReturnValue(request);
			jest.spyOn(target, 'req').mockReturnValue(request);
		});

		it('should call the resource with the given method and the configured timeout and number of retries', async () => {
			const result = await target.patch(
				'the resource that I want',
				{
					payload: 'value',
				},
				'my content type',
			);

			expectCallsLike(target.req, [
				Methods.PATCH,
				'the resource that I want',
				'my content type',
			]);
			expectCallsLike(request.send, [
				{
					payload: 'value',
				},
			]);
			expect(result).toEqual('expected result');
		});
	});

	describe(methods.put, () => {
		let request: superagent.SuperAgentRequest;

		beforeEach(() => {
			request = Promise.resolve({ body: 'expected result' }) as any;
			request.send = jest.fn().mockReturnValue(request);
			jest.spyOn(target, 'req').mockReturnValue(request);
		});

		it('should call the resource with the given method and the configured timeout and number of retries', async () => {
			const result = await target.put(
				'the resource that I want',
				{
					payload: 'value',
				},
				'my content type',
			);

			expectCallsLike(target.req, [
				Methods.PUT,
				'the resource that I want',
				'my content type',
			]);
			expectCallsLike(request.send, [
				{
					payload: 'value',
				},
			]);
			expect(result).toEqual('expected result');
		});
	});

	describe(methods.get, () => {
		let request: superagent.SuperAgentRequest;

		beforeEach(() => {
			request = Promise.resolve({ body: 'expected result' }) as any;
			request.query = jest.fn().mockReturnValue(request);
			jest.spyOn(target, 'req').mockReturnValue(request);
		});

		it('should call the resource with the given querystring the configured timeout and number of retries', async () => {
			const result = await target.get('the resource that I want', {
				payload: 'value',
			});

			expectCallsLike(target.req, [Methods.GET, 'the resource that I want']);
			expectCallsLike(request.query, [
				{
					payload: 'value',
				},
			]);
			expect(result).toEqual('expected result');
		});

		it('should call the resource with no querystring when it is not informed', async () => {
			const result = await target.get('the resource that I want');

			expectCallsLike(target.req, [Methods.GET, 'the resource that I want']);
			expectCallsLike(request.query);
			expect(result).toEqual('expected result');
		});
	});
});


test/unit/get-api-config-from-map.spec.ts
import { ApiConfig, getApiConfigFromMap } from '../../src';
import ms = require('ms');

describe(getApiConfigFromMap.name, () => {
	it('should return all the default values for the non informed envs vars', () => {
		const envs = new Map(
			Object.entries({
				MY_API_URL: 'my url',
			}),
		);

		const result = getApiConfigFromMap(envs, 'MY_API');

		expect(result).toEqual({
			timeout: ms('30s'),
			url: 'my url',
			retries: undefined,
			contentType: undefined,
			logRequests: false,
		} as ApiConfig);
	});

	it('should return all the informed default values for the non informed envs vars', () => {
		const envs = new Map(
			Object.entries({
				MY_API_BASE_URI: 'my url',
			}),
		);

		const result = getApiConfigFromMap(envs, 'MY_API', {
			contentType: 'application/pokemon',
			retries: 10,
			timeout: ms('3s'),
		});

		expect(result).toEqual({
			timeout: ms('3s'),
			url: 'my url',
			retries: 10,
			contentType: 'application/pokemon',
			logRequests: false,
		} as ApiConfig);
	});

	it('should return all the informed values', () => {
		const envs = new Map(
			Object.entries({
				MY_API_BASE_URL: 'my url',
				MY_API_TIMEOUT: '10s',
				MY_API_RETRIES: '5',
				MY_API_LOG_REQUESTS: 'true',
				MY_API_CONTENT_TYPE: 'application/he-man',
			}),
		);

		const result = getApiConfigFromMap(envs, 'MY_API', {
			retries: 10,
			timeout: ms('3s'),
		});

		expect(result).toEqual({
			timeout: ms('10s'),
			url: 'my url',
			retries: 5,
			contentType: 'application/he-man',
			logRequests: true,
		} as ApiConfig);
	});
});


src/get-api-config.ts
import { Defaults, ApiConfig } from './api-config';

const ms = require('ms');
const DEFAULT_TIMEOUT = ms('30s');

export function getMs(value: string | undefined): number {
	return value ? ms(value) : NaN;
}

export function getApiConfig(
	envs: Record<string, string | undefined>,
	apiName: string,
	defaults?: Defaults,
): ApiConfig {
	return {
		timeout:
			getMs(envs[`${apiName}_TIMEOUT`]) || defaults?.timeout || DEFAULT_TIMEOUT,
		retries: Number(envs[`${apiName}_RETRIES`]) || defaults?.retries,
		contentType: envs[`${apiName}_CONTENT_TYPE`] || defaults?.contentType,
		url:
			envs[`${apiName}_BASE_URL`] ||
			envs[`${apiName}_BASE_URI`] ||
			envs[`${apiName}_URL`] ||
			envs[`${apiName}_URI`]!,
		logRequests: envs[`${apiName}_LOG_REQUESTS`] === 'true',
	};
}

export function getApiConfigFromMap(
	envs: Map<string, string>,
	apiName: string,
	defaults?: Defaults,
): ApiConfig {
	return {
		timeout:
			getMs(envs.get(`${apiName}_TIMEOUT`)) ||
			defaults?.timeout ||
			DEFAULT_TIMEOUT,
		retries: Number(envs.get(`${apiName}_RETRIES`)) || defaults?.retries,
		contentType: envs.get(`${apiName}_CONTENT_TYPE`) || defaults?.contentType,
		url:
			envs.get(`${apiName}_BASE_URL`) ||
			envs.get(`${apiName}_BASE_URI`) ||
			envs.get(`${apiName}_URL`) ||
			envs.get(`${apiName}_URI`)!,
		logRequests: envs.get(`${apiName}_LOG_REQUESTS`) === 'true',
	};
}


src/base-http-client.ts
import request = require('superagent');
import { ApiConfig } from './api-config';
import { Logger } from './logger';
import Agent = require('agentkeepalive');

export enum Methods {
	DELETE = 'delete',
	GET = 'get',
	PATCH = 'patch',
	POST = 'post',
	PUT = 'put',
}

/**
 * A base http client with request logging capabilities and standardized entry config
 */
export class BaseHttpClient {
	private readonly defaultContentType: string;
	private readonly agent: Agent;
	/**
	 * Create a new instance of BaseHttpClient
	 * @param config the client configuration
	 * @param logger the logger to use in case of request logging being enabled
	 */
	constructor(private config: ApiConfig, private logger: Logger) {
		this.defaultContentType = this.config.contentType || 'application/json';
		this.agent = new (
			this.config.url.toLowerCase().startsWith('https:')
				? Agent.HttpsAgent
				: Agent
		)(
			config.agent || {
				maxFreeSockets: 50,
				timeout: 60000,
				freeSocketTimeout: 30000,
			},
		);
	}

	/**
	 * Mounts a request with the specified method
	 * @param method The method to use
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type set
	 */
	req(
		method: Methods,
		resource: string,
		contentType = this.defaultContentType,
	) {
		const result = request[method](`${this.config.url}/${resource}`)
			.agent(this.agent)
			.timeout(this.config.timeout)
			.retry(this.config.retries ?? 0)
			.set('content-type', contentType);

		return this.config.logRequests
			? result.use(require('superdebug')(this.logger.info.bind(this.logger)))
			: result;
	}

	/**
	 * Mounts a post request
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @parm payload The payload to be sent
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type and body set
	 */
	async post(resource: string, payload: object, contentType?: string) {
		return (await this.req(Methods.POST, resource, contentType).send(payload))
			.body;
	}

	/**
	 * Mounts a get request
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type and body set
	 */
	async get(resource: string, queryParams?: object) {
		const req = this.req(Methods.GET, resource);
		return (await (queryParams ? req.query(queryParams) : req)).body;
	}

	/**
	 * Mounts a put request
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @payload The payload to be sent
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type and body set
	 */
	async put(resource: string, payload: object, contentType?: string) {
		return (await this.req(Methods.PUT, resource, contentType).send(payload))
			.body;
	}

	/**
	 * Mounts a patch request
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @payload The payload to be sent
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type and body set
	 */
	async patch(resource: string, payload: object, contentType?: string) {
		return (await this.req(Methods.PATCH, resource, contentType).send(payload))
			.body;
	}

	/**
	 * Mounts a delete request
	 * @param resource the resource to be accessed (the base url is used from the config)
	 * @payload The payload to be sent
	 * @param contentType the contentType to use (default is get from config or, if not informed, application/json)
	 * @returns an superagent request object with timeout, retry and content-type and body set
	 */
	async delete(resource: string, contentType?: string) {
		return (await this.req(Methods.DELETE, resource, contentType)).body;
	}
}
